### PR TITLE
do not create empty userid when attribute does not have allowed chars

### PR DESCRIPTION
**Pull request by blizzz**
If some funny attribute contains only characters that are not allowed for user ids, one empty userid would have been generated. Comes with unit test.

### CODE REVIEW

File: apps/user_ldap/lib/Access.php [hunk 1]
```diff
@@ -579,7 +579,19 @@ public function dn2ocname($fdn, $ldapName = null, $isUser = true, &$newlyMapped
...
-			$intName = $this->sanitizeUsername($username);
+			try {
+				$intName = $this->sanitizeUsername($username);
+			} catch (\InvalidArgumentException $e) {
+				\OC::$server->getLogger()->logException($e, [
+					'app' => 'user_ldap',
+					'level' => Util::WARN,
+				]);
+				// we don't attempt to set a username here. We can go for
+				// for an alternativ 4 digit random number as we would append
```
Comment (MorrisJobke) [inline apps/user_ldap/lib/Access.php]: `alternative` ;)

File: apps/user_ldap/lib/Access.php [hunk 2]
```diff
@@ -1300,14 +1312,18 @@ public function sanitizeUsername($name) {
...
+		$name = iconv('UTF-8', 'ASCII//TRANSLIT', trim($name));
 
 		// Replacements
 		$name = str_replace(' ', '_', $name);
 
 		// Every remaining disallowed characters will be removed
 		$name = preg_replace('/[^a-zA-Z0-9_.@-]/u', '', $name);
 
+		if($name === '') {
+			throw new \InvalidArgumentException('provided name template for username does not contain any allowed characters');
```
Comment (MorrisJobke) [inline apps/user_ldap/lib/Access.php]: Missing in PHPDoc

**Review by MorrisJobke — APPROVED** [review]
Beside the little nitpicks: 👍

**Comment by MorrisJobke** [comment]
unit tests fail `iconv(): Detected an illegal character in input string` 🤔

File: apps/user_ldap/lib/Access.php (hunk 2 above)
Comment (blizzz) [reply apps/user_ldap/lib/Access.php]: mh, PhpStorm not complaining, usually it does… odd. Anyway you're right.

**Comment by blizzz** [comment]
> unit tests fail iconv(): Detected an illegal character in input string thinking

iconv :-/

> »Caution
>
> If and how //TRANSLIT works exactly depends on the system's iconv() implementation (cf. ICONV_IMPL). Some implementations are known to ignore //TRANSLIT, so the conversion is likely to fail for characters which are illegal for the out_charset.« 
[php doc](https://secure.php.net/manual/en/function.iconv.php)

Yet there are no runtime or any config options available for PHP. So I changed the behaviour slightly: try to transliterate, be silent about issues, otherwise non-ascii characters are removed anyhow.

**Review by MorrisJobke — APPROVED** [review]

**Review by rullzer — APPROVED** [review]

**Pull request merged**